bin/gobblin-mapreduce.sh

```diff
@@ -14,26 +14,46 @@ fi
 
 JOB_TRACKER_URL=$1
 FS_URL=$2
-JOB_CONFIG_FILE=$3
+JOB_CONFIG_FILE=$4
+
+# Convert job jar's relative path to absolute path
+DIRNAME=`dirname "$3"`
+JARNAME=`basename "$3"`
+DIRNAME=`cd "$DIRNAME">/dev/null; pwd`
+JOB_JAR_FILE="$DIRNAME/$JARNAME"
 
 FWDIR="$(cd `dirname $0`/..; pwd)"
+FWDIR_LIB=$FWDIR/lib
 
 # Jars Gobblin runtime depends on
-LIBJARS=$FWDIR/lib/common-api.jar,$FWDIR/lib/metastore.jar,$FWDIR/lib/qualitychecker.jar,\
-$FWDIR/lib/source.jar,$FWDIR/lib/converter.jar,$FWDIR/lib/writer.jar,$FWDIR/lib/utility.jar,\
-$FWDIR/lib/guava-15.0.jar,$FWDIR/lib/avro-1.7.1.jar,$FWDIR/lib/avro-mapred-1.7.6.jar,\
-$FWDIR/lib/metrics-core-3.0.2.jar,$FWDIR/lib/gson-2.2.4.jar
-
-export HADOOP_CLASSPATH=$FWDIR/lib/*
+LIBJARS=$FWDIR_LIB/gobblin-metastore.jar,$FWDIR_LIB/gobblin-metrics.jar,\
+$FWDIR_LIB/gobblin-core.jar,$FWDIR_LIB/gobblin-api.jar,$FWDIR_LIB/gobblin-utility.jar,\
+$FWDIR_LIB/guava-15.0.jar,$FWDIR_LIB/avro-1.7.6.jar,$FWDIR_LIB/avro-mapred-1.7.6.jar,\
+$FWDIR_LIB/metrics-core-3.0.2.jar,$FWDIR_LIB/gson-2.2.4.jar,$FWDIR_LIB/joda-time-1.6.jar,$FWDIR_LIB/data-1.15.9.jar,\
+$JOB_JAR_FILE
+
+# Add libraries to the Hadoop classpath
+GOBBLIN_DEP_JARS=""
+for jarFile in `ls $FWDIR_LIB/*`
+do
+  GOBBLIN_DEP_JARS=${GOBBLIN_DEP_JARS}:$jarFile
+done
+
+# Honor Gobblin dependencies
+export HADOOP_USER_CLASSPATH_FIRST=true
```

Comment:
Good catch! According to HADOOP-9450 and the fact that Hadoop 1.2.1 has "mapreduce.user.classpath.first" which has been renamed to "mapreduce.job.user.classpath.first" in Hadoop 2, it seems to me that loading the Gobblin related jars first is not a must in the Hadoop 1 environment (at least in 1.2.1). I used these props in Hadoop 2 (2.3 and 2.6 HDP distribution where not loading the Gobblin related dependencies first will lead to runtime issues)

As for the mapreduce properties:
It can be misleading that the script wants to use a Hadoop 2 properties even when the job is submitted to Hadoop1. MAPREDUCE-4521 re-added the old property as deprecated to Hadoop 2. So it can be used in the Hadoop 2 environment. Probably both of these mapreduce properties could be used together to play safe? 
Or another option would be to check the Hadoop version in the script and set properties accordingly (e.g: probably skip the whole procedure in Hadoop 1). Similar check is done
in Pig, e.g:
HADOOP_CORE_JAR=`echo ${HADOOP_HOME}/hadoop-core*.jar`
if [ -z "$HADOOP_CORE_JAR" ]; then
    HADOOP_VERSION=2
else
    HADOOP_VERSION=1
fi
...
What do you think?